### PR TITLE
Add new crypto callback for RSA with padding.

### PR DESCRIPTION
**Pull request by ColtonWilley**
# Description

Adds a new crypto callback for RSA with padding. The new callback optionally handles RSA padding itself, while maintaining existing fallback models.

### CODE REVIEW

**Comment by ColtonWilley** [comment]
Retest this please.

File: tests/api.c [hunk 1]
```diff
@@ -83658,6 +83658,13 @@ static int test_CryptoCb_Func(int thisDevId, wc_CryptoInfo* info, void* ctx)
                 info->pk.rsa.type, ret, *info->pk.rsa.outLen);
         #endif
         }
+        #ifdef WOLF_CRYPTO_CB_RSA_PAD
+        else if (info->pk.type == WC_PK_TYPE_RSA_PKCS ||
+                 info->pk.type == WC_PK_TYPE_RSA_PSS  ||
+                 info->pk.type == WC_PK_TYPE_RSA_OAEP) {
+            ret = CRYPTOCB_UNAVAILABLE; /* fallback to software */
```
Comment (SparkiDev) [inline tests/api.c:83665]: I would like to see a software implementation of this.
We can expose APIs in rsa.c to help make this work.

File: wolfcrypt/src/wc_pkcs11.c [hunk 2]
```diff
@@ -1809,6 +1809,63 @@ static int Pkcs11RsaPrivateKey(Pkcs11Session* session, RsaKey* rsaKey,
     return ret;
 }
 
+int wc_hash2sz(int hType)
```
Comment (SparkiDev) [inline wolfcrypt/src/wc_pkcs11.c:1812]: Please add function block comments like the other functions in this file.

File: wolfcrypt/src/wc_pkcs11.c [hunk 3]
```diff
@@ -1809,6 +1809,63 @@ static int Pkcs11RsaPrivateKey(Pkcs11Session* session, RsaKey* rsaKey,
...
+    switch(hType) {
+    case WC_HASH_TYPE_SHA:
+        return 20;
+    case WC_HASH_TYPE_SHA256:
+        return 32;
+    case WC_HASH_TYPE_SHA384:
+        return 48;
+    case WC_HASH_TYPE_SHA512:
+        return 64;
+    case WC_HASH_TYPE_SHA224:
```
Comment (SparkiDev) [inline wolfcrypt/src/wc_pkcs11.c:1823]: Put SHA224 above SHA256 for the sake of easy reading.

**Review by SparkiDev — CHANGES_REQUESTED** [review]
Has this been tested with a PKCS#11 implementation?

**Comment by ColtonWilley** [comment]
Yes, this has been tested against softHSM.

File: tests/api.c (hunk 1 above)
Comment (ColtonWilley) [reply tests/api.c:83665]: I implemented a software leg for the RSA private key operations similar to how the existing callback for testing operates. I was able to do it with existing APIs, let me know if that is what you had in mind.

**Review by SparkiDev — APPROVED** [review]

**Comment by dgarske** [comment]
Retest this please. History lost, too old

**Comment by ColtonWilley** [comment]
@SparkiDev I have added a PSS example to the PKCS11 examples, the rest of the code did not need to be modified.

https://github.com/wolfSSL/wolfssl-examples/pull/456

**Review by dgarske — APPROVED** [review]
Over to @SparkiDev

**Review by SparkiDev — APPROVED** [review]

**Pull request merged**